### PR TITLE
Add ambient autoenroll mechanism

### PR DESCRIPTION
**Pull request by bleggett**
**Please provide a description of this PR:**

Implements parts of https://github.com/istio/istio/issues/53578 (the autoenroll by default parts)

## New flags

- Adds a new Helm value to the CNI chart - `ambient.autoEnroll: true|false`
- Adds a new Helm value to the CNI chart - `ambient.autoEnrollExcludeNamespaces: [list]`

## Explanation
If `ambient.enabled` is true, and `ambient.autoEnroll` is true, Ambient will switch from default opt-in, to default opt-out.

Note that default opt-out still respects all the existing `istio.io/dataplane-mode` namespace/pod combinatorics and precedence - those still always "win", but autoEnroll will grab all the stuff that isn't already explicitly labeled one way or the other.

This means that, for instance, `istio-cni` will try to autoenroll every workload on its node that is not

- in a namespace in the `autoEnrollExcludeNamespaces` denylist.
- is not already explicitly excluded with the existing `istio.io/dataplane-mode=none` on _either_ the namespace or the pod.

This **_also_** means that, for instance, `istio-cni` will behave like it did before, respecting explicit labels, and enroll a workload on its node that:
- is in a namespace in the `autoEnrollExcludeNamespaces` denylist.
- is **_explicitly_** included in the mesh with a `istio.io/dataplane-mode=ambient` label.

(this latter case is useful if you want to put `istio-system` in `autoEnrollExcludeNamespaces` so _everything_ in that namespacee isn't grabbed, but you still want to selectively opt-in one or two workloads running in that namespace).

Short version: Labels still work just like they did before, and always "beat" autoenroll, if present - but autoenroll will grab pods that are not otherwise labeled or excluded.

## Side note
`istio-cni` has a preexisting `excludedNamespaces` flag, which is purely for sidecar mode, and was until now ignored in the ambient watcher. I originally tried to overload this so we could use it for both, but the semantics of this do not work in practice so that is still entirely separate/only for sidecar/unused in ambient.

## What this does not do

https://github.com/istio/istio/issues/53578 mentions another requirement, which is I think entirely separate - being able to enforce that all workloads are enrolled in the mesh, as a centralized config.

Unfortunately, our existing opt-in model has an explicit behavior that [pod-level opt-out labels always beat namespace opt-in labels](https://istio.io/latest/docs/ambient/usage/add-workloads/#pod-selection-logic-for-ambient-and-sidecar-modes) - and implementing this kind of enforcement effectively would require we break/change that behavior.

There are a few ways to do this, which might confuse users more or less:

- We could do something like "in autoEnroll mode, pod-level opt-outs are _categorically ignored_, and only namespace-level opt-outs are allowed". This is a bit awkward for users tho, as I can see people wanting both autoEnroll, and pod-level opt-outs, and not liking that they are mutually exclusive.

- We could introduce a new namespace-level config/label/whatever, that disables pod-level opt-outs. Something like `ambient.istio.io/disableOptOuts`. This has the benefit of being selective (maybe some namespaces should be stricter than others on this point), and works regardless of whether or not autoEnroll is in use.

- We could do neither and let people use whatever policy enforcement tools (OPA/CEL admission webhooks, whatever) they already have in their cluster to block pods that have opt-out labels.

It's worth chatting about how we want to do this, but this PR is not solving for that second requirement, because that's changing our existing label semantics - which this is not doing.

### CODE REVIEW

File: manifests/charts/istio-cni/values.yaml [hunk 1]
```diff
@@ -53,7 +54,16 @@ _internal_defaults_do_not_set:
     dnsCapture: false
     # If enabled, and ambient is enabled, enables ipv6 support
     ipv6: true
-
+    # If enabled, and ambient is enabled, will switch from default opt-in, to default opt-out.
+    # This will auto-enroll pods in *all* namespaces if no dataplane labels are present on either pod or namespace.
+    # Pods in `autoEnrollExcludeNamespaces` will not be autoenrolled.
+    # Explicit pod and namespace enrollment labels will still be respected in this mode.
+    autoEnroll: false
```
Comment (hzxuzhonghu) [inline manifests/charts/istio-cni/values.yaml:61]: wow, this is not a good knob. Is that just so simple onece for all, i donot think so, users need to exclude more with this.

Comment (bleggett) [reply manifests/charts/istio-cni/values.yaml:61]: Yep, and they can with the existing labels - this isn't mutually exclusive with those.

What are some scenarios here you're thinking of, that aren't handled?

Comment (hzxuzhonghu) [reply manifests/charts/istio-cni/values.yaml:61]: I am skeptical how useful this is?  Most users won't turn on this,  even they donot enable sidecar injection for all the workload.

Comment (ilrudie) [reply manifests/charts/istio-cni/values.yaml:61]: I'm not sure it tracks to assume ambient adoption patterns with be identical to sidecar adoption. The sidecar approach is inherently more costly, more disruptive and causes immediate behavioral changes. Ambient does a pretty good job of reducing or eliminating those problems, which to my mind makes adopting an "opt-out" model way more approachable.

Comment (bleggett) [reply manifests/charts/istio-cni/values.yaml:61]: Yeah - I don't think the sidecar comparison makes sense. Ambient already supports enrolling workloads with just a namespace label. This is a minor step beyond that.

**Comment by howardjohn** [comment]
What i had envisioned is a matchExpression on namespace+pod labels, just
like webhooks (and a few Istio features like DiscoverySelectors, and some
other things we have been discussing). i haven't had a chance to think this
PR through too much (kubecon) but would be good to compare that approach

**Comment by bleggett** [comment]
> What i had envisioned is a matchExpression on namespace+pod labels, just like webhooks (and a few Istio features like DiscoverySelectors, and some other things we have been discussing). i haven't had a chance to think this PR through too much (kubecon) but would be good to compare that approach

If we did do something like that (I am not opposed to it), it should be a proper API/CRD IMO.

**Comment by bleggett** [comment]
Actually - unrelated to the API discussion, I am reminded that https://github.com/istio/istio/issues/49009 is effectively a blocker for this.

Since we do not unhook pods on `istio-cni` uninstall or upgrade (we can't tell the difference), that will interact really poorly for people that use autoenroll.

The only way they could "fix" their workloads would be to explicitly disable everything with labels, and then uninstall, which is not really acceptable.

**Comment by bleggett** [comment]
Closing this for now, the selector-based API discussed here will likely be repurposed/employed for an iteration on the DNS exclusion stuff, in order to address concerns people had in the Nov. 20 Ambient WG about the security implications of pod-level opt-out: https://github.com/istio/api/pull/3361

**Comment by therealmitchconnors** [comment]
@bleggett now that #49009  is closed, can AutoEnroll proceed safely?

**Comment by bleggett** [comment]
> @bleggett now that #49009 is closed, can AutoEnroll proceed safely?

Yeah, closing that resolved my concerns with this (and a few other longstanding issues to boot) - it will probably be a bit before I circle back to this `autoEnroll` feature tho, so just closing this for now.

**Pull request closed without merging**